rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,128 @@
+
+Because only a single version of the client tools will be cached at a time, we do not need to worry about cleaning up the cache.
+
+### Package upgrade
+
+As an alternative to caching, we've considered directly upgrading the Teleport packages.
+
+This approach will be difficult to implement because we support multiple package repositories and installation methods. The amount of maintenance and testing required to support this solution is probably not worth the effort at this time.
+
+Another downside to this approach would be that upgrades would require sudo/admin privileges.
+
+### Config
+
+Automatic updates for client tools can be configured through tsh configuration. If `DISABLE_AUTO_UPDATE=true`, then auto updates for client tools will be disabled. If `DISABLE_UPDATE_PROMPT=true`, then the user will not be prompted to confirm the update. Both values will default to false.
+
+### Binary Execution
+
+The client tools will need a way to execute the cached binaries. When tsh/tctl is executed, it will now need to decide whether to execute the command as usual, or fork a new process using the cached binary.
+
+One way to handle execution of cached binaries is to use the os.Executable function. os.Executable give us the path of the binary executable. The flow might look something like:
+- If executing `tsh login` execute as usual (cached tsh binary will not support TouchID on macOS)
+- If executable is in cache directory (.tsh/bin/tsh), then execute command as usual
+- Else try to execute tsh from cache if exists
+
+An alternative method could be use to a flag like `--no-cache` to determine if a cached version should be executed. The flow might look something like:
+- If `--no-cache=true`, execute as usual
+- If `--no-cache=false`, try to execute tsh from cache with `--no-cache=true`
+- If tsh does not exist in cache, execute as usual
+
+Another option could be to have users update their `PATH` env with `export PATH=$HOME/.tsh/bin:$PATH`. Adding the tsh bin directory to the `PATH` environment in the user's `.profile` can be handled by tsh. Users would then execute the cached binary directly. This seems like a simpler approach than the above methods. The problem is that the cached binary would not support TouchID for macOS users.
+
+I believe macOS software needs to be notarized as a requirement for TouchID. For this reason Teleport provides a separate signed tsh download at `https://cdn.teleport.dev/tsh-13.3.4.pkg`. We could change the download method for macOS users, but this might not be very straight forward to implement.
```

Comment:
There's no reason for us to ship unsigned binaries. We will be fixing this this quarter.